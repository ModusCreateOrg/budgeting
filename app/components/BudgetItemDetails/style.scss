@import 'theme/variables';

.budgetItemDetails {

  .budgetItemDetailsTitle {
    font-size: 15px;
    font-family: $font-default;
  }

  .budgetItemDetailsSubtitle {
    font-family: $font-default;

    &.neg {
      color: $red;
    }

    &.pos {
      color: $green;
    }
  }

  @media only screen and (max-width: $screen-small) {
    margin: 0 auto;
  }

}
