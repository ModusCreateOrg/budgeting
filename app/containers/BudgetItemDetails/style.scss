@import 'theme/variables';

.budgetItemDetails {

  .goBackButton {
    display: inline-block;
    border: solid 1px $green;
    border-radius: 3px;
    padding: 0.5rem 1.5rem;
    font-size: 16px;
    cursor: pointer;
    box-shadow: 1px 1px 10px rgba(0, 0, 0, 0.1);

    &:hover {
      box-shadow: 1px 1px 10px rgba(0, 0, 0, 0.15);
      opacity: 0.6;
    }

    &:active {
      box-shadow: 1px 1px 10px rgba(0, 0, 0, 0.2);
      opacity: 1;
    }
  }

  .textGreen {
    color: $green;
  }

  .textRed {
    color: $red;
  }

}
