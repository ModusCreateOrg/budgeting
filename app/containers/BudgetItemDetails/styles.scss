.inflowBalance {
  color: #00ff00;
}

.outflowBalance{
  color: #ff0000;
}

.backButton{
  margin-bottom: 1rem;
}
