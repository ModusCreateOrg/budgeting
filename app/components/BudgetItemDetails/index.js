// @flow
import * as React from 'react';
import cx from 'classnames';
import { formatPercentageAmount } from 'utils/formatAmount';
import styles from './style.scss';

type BudgetItemProps = {
  transaction: Object,
  inflowBalance: number,
  outflowBalance: number,
};

class BudgetItemDetails extends React.Component<BudgetItemProps> {
  render() {
    const { transaction, inflowBalance, outflowBalance } = this.props;
    if (!transaction) {
      return <span>Could not find the corresponding budget item</span>;
    }

    const percentageAmount = formatPercentageAmount(transaction.value, inflowBalance, outflowBalance);

    const subtitleClassname = cx([styles.budgetItemDetailsSubtitle], {
      [styles.neg]: percentageAmount.isNegative,
      [styles.pos]: !percentageAmount.isNegative,
    });

    return (
      <div className={styles.budgetItemDetails}>
        <h1 className={styles.budgetItemDetailsTitle}>{transaction.description}</h1>
        <p className={subtitleClassname}>{percentageAmount.text}</p>

      </div>
    );
  }
}
export default BudgetItemDetails;
