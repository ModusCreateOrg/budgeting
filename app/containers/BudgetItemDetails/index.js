// @flow
import * as React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { injectAsyncReducers } from 'store';
import { withRouter } from 'react-router-dom';
import type { Match, RouterHistory } from 'react-router-dom';
import { getTransactions, getInflowBalance, getOutflowBalance } from '../../selectors/transactions';
import transactionReducer from '../../modules/transactions';
import type { Transaction } from '../../modules/transactions';
import formatPercentage from '../../utils/formatPercentage';
import formatAmount from '../../utils/formatAmount';
import styles from './style.scss';

// inject reducers that might not have been originally there
injectAsyncReducers({
  transactions: transactionReducer,
});

type BudgetItemDetailsProps = {
  match: Match,
  transactions: Transaction[],
  history: RouterHistory,
  inflowBalance: number,
  outflowBalance: number,
};

export const BudgetItemDetails = ({
  match,
  transactions,
  history,
  inflowBalance,
  outflowBalance,
}: BudgetItemDetailsProps) => {
  const getItem = itemIdToGet => ({ id }) => id === Number(itemIdToGet);
  const { itemId } = match.params;
  const item = transactions.filter(getItem(itemId))[0];

  const isInflow = item.value > 0;
  const percentageOfTotal = formatPercentage(item.value / (isInflow ? inflowBalance : outflowBalance) * 100, {
    decimals: 1,
  });

  return (
    <div className={styles.budgetItemDetails}>
      <div className={`goBackButton ${styles.goBackButton}`} onClick={history.goBack} role="button" tabIndex="0">
        Back
      </div>
      <div>
        <h1>{item.description}</h1>
        <h2>
          Amount {isInflow ? 'in' : 'spent'}: {' '}
          {isInflow ? (
            <span className={styles.textGreen}>{formatAmount(item.value).text}</span>
          ) : (
            <span className={styles.textRed}>{formatAmount(item.value).text}</span>
          )}
        </h2>
        <h3>
          which is {percentageOfTotal} of total {isInflow ? 'inflow' : 'outflow'} {' '}
          {isInflow ? (
            <span className={styles.textGreen}>{formatAmount(inflowBalance).text}</span>
          ) : (
            <span className={styles.textRed}>{formatAmount(outflowBalance).text}</span>
          )}
        </h3>
      </div>
    </div>
  );
};

BudgetItemDetails.propTypes = {
  inflowBalance: PropTypes.number.isRequired,
  outflowBalance: PropTypes.number.isRequired,
  history: PropTypes.shape({
    goBack: PropTypes.func.isRequired,
  }).isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      itemId: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
  transactions: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      description: PropTypes.string.isRequired,
      value: PropTypes.number.isRequired,
    })
  ).isRequired,
};

const mapStateToProps = state => ({
  transactions: getTransactions(state),
  inflowBalance: getInflowBalance(state),
  outflowBalance: getOutflowBalance(state),
});
const mapDispatchToProps = () => ({});

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(BudgetItemDetails));
