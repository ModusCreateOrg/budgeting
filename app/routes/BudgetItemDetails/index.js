// @flow
import React, { Component } from 'react';
import Chunk from 'components/Chunk';

const loadBudgetItemContainer = () =>
  import('containers/BudgetItemDetails' /* webpackChunkName: "budgetItemDetails" */);

class BudgetItemDetails extends Component<{}> {
  render() {
    return <Chunk load={loadBudgetItemContainer} />;
  }
}

export default BudgetItemDetails;
