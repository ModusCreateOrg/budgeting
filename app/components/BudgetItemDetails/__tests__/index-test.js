import React from 'react';
import renderer from 'react-test-renderer';
import BudgetItemDetails from '../';

// mock nested component
jest.mock('components/PieChart');

it('renders correctly when transaction exists with negative value', () => {
  const transaction = { value: -10, description: 'simple transaction with negative value', categoryId: 1 };
  const tree = renderer
    .create(<BudgetItemDetails transaction={transaction} inflowBalance={100} outflowBalance={50} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly when transaction exists with positive value', () => {
  const transaction = { value: 10, description: 'simple transaction with positive value', categoryId: 1 };
  const tree = renderer
    .create(<BudgetItemDetails transaction={transaction} inflowBalance={100} outflowBalance={50} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly when transaction does not exists', () => {
  const tree = renderer.create(<BudgetItemDetails />).toJSON();
  expect(tree).toMatchSnapshot();
});
