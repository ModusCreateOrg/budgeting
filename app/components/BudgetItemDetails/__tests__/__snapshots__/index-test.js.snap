// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when transaction does not exists 1`] = `
<span>
  Could not find the corresponding budget item
</span>
`;

exports[`renders correctly when transaction exists with negative value 1`] = `
<div
  className="budgetItemDetails"
>
  <h1
    className="budgetItemDetailsTitle"
  >
    simple transaction with negative value
  </h1>
  <p
    className="budgetItemDetailsSubtitle neg"
  >
    -20.00% of $50.00 (outflow)
  </p>
</div>
`;

exports[`renders correctly when transaction exists with positive value 1`] = `
<div
  className="budgetItemDetails"
>
  <h1
    className="budgetItemDetailsTitle"
  >
    simple transaction with positive value
  </h1>
  <p
    className="budgetItemDetailsSubtitle pos"
  >
    +10.00% of $100.00 (inflow)
  </p>
</div>
`;
